### ISSUE TITLE
catboost does not use all cores

### ISSUE THREAD

**Issue by kdlin** [catboost/catboost#1256]
Problem: catboost uses 10 or 11 cores when 16 cores are available
catboost version: 0.22
Operating System: Linux 4.4.0-1100-aws x86_64
CPU: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
# GPU:

I have a large dataset. When I use LightGBM, it uses all cores. Then I switched to CatBoost and found not all cores are used. According to the CatBoost documentation, the default is using all cores. Please advise on how I can debug.

**Comment by kizill** [comment]
Hello!
Could you provide additional info about dataset properties - numer of float, categorical and text features and objects count in train and validation datasets? Also we need info about training params: loss function, custom metrics etc.

**Comment by kdlin** [comment]
The training dataset has 1.2M rows and 218 columns. Among these 218 columns, 42 are categorical features. The rest are floats or integers, no text features. The validation dataset has 120K rows.

The loss function and metric are both logloss. No custom metrics are used. The rest of the parameters are:
        l2_leaf_reg= between 0 and 2
        max_depth = between 2 and 10
        colsample_bylevel = between 0.5 and 1.0
        iterations=850,
        eval_metric='Logloss',
        random_seed=42,
        verbose=False,
        loss_function='Logloss',
        od_type = 'Iter',
        od_wait = 35,
        use_best_model = True

**Comment by kdlin** [comment]
@kizill I'd like to give you two more examples to illustrate the problems.
(1) Compared to LightGBM, CatBoost does not use CPU effectively. 
(2) With GPU, it uses a huge amount of memory in CPU's RAM.

The dataset used is Census Income dataset. A cleanup version for using it for classification can be obtained via the CHAP package, by the following command: (after you install the shap package)
df, y_data=shap.datasets.adult()

I am using an AWS g3 instance. I run both CatBoost and LightGBM at the same time, along with a Bayesian optimizer, Optuna. The following is a screen shot of top command. 
PID 8180: LightGBM with Optuna
PID 11738: CatBoost with Optunia
![Screen Shot 2020-05-03 at 2 00 12 PM](https://user-images.githubusercontent.com/25801835/80925684-bf019300-8d46-11ea-85e7-114d272c987b.png)
As you can use, LightGBM is capable of using more cores. A snip of the CatBoost code can be seen below:
![Screen Shot 2020-05-03 at 2 04 12 PM](https://user-images.githubusercontent.com/25801835/80925726-0be56980-8d47-11ea-8914-347cd57bdc6d.png)

(2) A problem with CatBoost in GPU mode is that it seems to use a huge amount of RAM. The same CatBoost code is now run in GPU, after specifying task_type = GPU. Now PID 12147 is from CatBoost and Optuna. As you can see, 135GB of RAM is virtual! 
![Screen Shot 2020-05-03 at 2 06 40 PM](https://user-images.githubusercontent.com/25801835/80925813-b1004200-8d47-11ea-98c7-a3cf31f2c8ee.png)

It does not use GPU effectively either. As you can see from the below screenshot from nvtop, the GPU efficiency can be improved as well.
![Screen Shot 2020-05-03 at 2 10 40 PM](https://user-images.githubusercontent.com/25801835/80925843-089ead80-8d48-11ea-846e-7c097546a9d3.png)

As a result, it also runs very slow in the GPU mode. Any pointers?

**Comment by kzamkov** [comment]
@dmsivkov

**Comment by dmsivkov** [comment]
@kdlin For (1) Can you test on the system without LightGBM or set explicit affinity mask for CatBoost?

and if possible collect VTune Threading analysis https://software.intel.com/content/www/us/en/develop/documentation/vtune-help/top/analyze-performance/parallelism-analysis-group/threading-analysis.html ?

**Comment by pgagarinov** [comment]
I have similar problem: 
<img width="1869" alt="image" src="https://user-images.githubusercontent.com/4868370/114790519-9030d100-9d8d-11eb-8cc3-62929becff5b.png">

The following command
```
model = CatBoostRegressor(..., num_threads=32)
model.fit(...)
```
uses exactly the half  of the dual CPU machine with the following specs:

<img width="754" alt="image" src="https://user-images.githubusercontent.com/4868370/114790689-db4ae400-9d8d-11eb-9bda-e63c1475c6be.png">

Can it be somehow related to having 2 Numa nodes?
